pipeline-adapters/mlrun-pipelines-kfp-v1-8/src/mlrun_pipelines/utils.py

```diff
@@ -12,11 +12,91 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+import os
 import tempfile
 import typing
 
-from mlrun_pipelines.helpers import new_pipe_metadata
-from mlrun_pipelines.imports import compiler, kfp  # noqa: F401
+import mlrun_pipelines.common.models
+import mlrun_pipelines.helpers
+import mlrun_pipelines.imports
+
+
+class ExtendedKfpClient(mlrun_pipelines.imports.Client):
+    def retry_run(
+        self,
+        run_id: str,
+    ) -> str:
+        """
+        Retries a given run by its run ID. If the run is not in a valid state for retry,
+        it creates a new run with the same pipeline and parameters.
+
+        :param run_id: The ID of the run to retry.
+        :type run_id: str
+        :raises ApiException: If the API request fails during the retry or new run creation process.
+        :raises ValueError: If the experiment ID cannot be found for the given run ID, or if
+                            the original run does not contain a valid pipeline specification.
+        :raises FileNotFoundError: If a temporary file for the workflow manifest cannot be created or accessed.
+        :return: The ID of the new or retried run.
+        :rtype: str
+        """
+        # Fetch run details
+        run_details = self.get_run(run_id).run
+        run_status = run_details.status
+
+        # Extract experiment ID from resource_references
+        experiment_id = next(
+            (
+                ref.key.id
+                for ref in run_details.resource_references
+                if ref.key.type == "EXPERIMENT"
+            ),
+            None,
+        )
+        if not experiment_id:
+            raise ValueError(f"Experiment ID not found for run ID: {run_id}")
+
+        valid_states_for_retry = {
+            mlrun_pipelines.common.models.RunStatuses.failed,
+            mlrun_pipelines.common.models.RunStatuses.error,
+        }
```

Comment:
In your HLD, you also mention the `Succeeded` and `Skipped` statuses
https://iguazio.atlassian.net/wiki/spaces/MLRUN/pages/183534076/Retry+Workflow+Feature+HLD#:~:text=Valid%20States%20for%20Retry%3A

Reply:
Suceeded and Skipped can be retried, but in those cases the a new workflow will be created as a clone of the previous one.